src/compiler/sys/in-memory-fs.ts

```diff
@@ -2,76 +2,225 @@ import type * as d from '../../declarations';
+  isDirectory: boolean;
+  size: number;
+  mtimeMs: number;
+  exists: boolean;
+  queueCopyFileToDest: string;
+  queueWriteToDisk: boolean;
+  queueDeleteFromDisk?: boolean;
+  useCache: boolean;
+}
+
+/**
+ * Storage format for the in-memory cache used to proxy the OS filesystem.
+ *
+ * Filesystem paths (of type `string`) are mapped to objects satisfying the
+ * `FsItem` interface.
+ */
+export type FsItems = Map<string, FsItem>;
+
+/**
+ * Options supported by write methods on the in-memory filesystem.
+ */
+export interface FsWriteOptions {
+  inMemoryOnly?: boolean;
+  clearFileCache?: boolean;
+  immediateWrite?: boolean;
+  useCache?: boolean;
+  /**
+   * An optional tag for the current output target for which this file is being
+   * written.
+   */
+  outputTargetType?: string;
+}
+
+/**
+ * Results from a write operation on the in-memory filesystem.
+ */
+export interface FsWriteResults {
+  changedContent: boolean;
+  queuedWrite: boolean;
+  ignored: boolean;
+}
+
+/**
+ * Options supported by read methods on the in-memory filesystem.
+ */
+export interface FsReadOptions {
+  useCache?: boolean;
+  setHash?: boolean;
+}
+
+/**
+ * Options supported by the readdir option on the in-memory filesystem.
+ */
+interface FsReaddirOptions {
+  inMemoryOnly?: boolean;
+  recursive?: boolean;
+  /**
+   * Directory names to exclude. Just the basename,
+   * not the entire path. Basically for "node_modules".
+   */
+  excludeDirNames?: string[];
+  /**
+   * Extensions we know we can avoid. Each extension
+   * should include the `.` so that we can test for both
+   * `.d.ts.` and `.ts`. If `excludeExtensions` isn't provided it
+   * doesn't try to exclude anything. This only checks against
+   * the filename, not directory names when recursive.
+   */
+  excludeExtensions?: string[];
+}
+
+/**
+ * A result from a directory read operation
+ */
+interface FsReaddirItem {
+  absPath: string;
+  relPath: string;
+  isDirectory: boolean;
+  isFile: boolean;
+}
+
+/**
+ * Information about a file in the in-memory filesystem.
+ */
+interface FsStat {
+  exists: boolean;
+  isFile: boolean;
+  isDirectory: boolean;
+  size: number;
+}
+
+/**
+ * Create an in-memory FS which proxies the underlying OS filesystem using an
+ * in-memory cache. FS writes can accumulate on the in-memory system, using an
+ * API similar to Node.js' `"fs"` module, and then be committed to disk as a
+ * unit.
+ *
+ * Files written to the in-memory system can be edited, deleted, and so on.
+ * This allows the compiler to proceed freely as if it is modifying the
+ * filesystem, modifying the world in whatever way suits it, while deferring
+ * actual FS writes until the end of the compilation process, making actual
+ * changes to the filesystem on disk contingent on an error-free build or any
+ * other condition.
+ *
+ * @param sys a compiler system object
+ * @returns an in-memory filesystem interface
+ */
 export const createInMemoryFs = (sys: d.CompilerSystem) => {
```

Comment:
As always, and perhaps especially with this file, these comments reflect what I learned from reading the code, looking for call sites across the codebase, and a little bit of github / slack spelunking here and there. So there very well may be inaccuracies!